README.md
<!--
SPDX-FileCopyrightText: 2022 Stephan Druskat <pyswh@sdruskat.net>

SPDX-License-Identifier: CC-BY-4.0
-->

# *pyswh* - a Python wrapper library for the Software Heritage API

*pyswh* aims to wrap interactions with the [Software Heritage REST API](https://archive.softwareheritage.org/api/1/) into a comfortable Python API.

[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=sdruskat_pyswh&metric=alert_status)](https://sonarcloud.io/summary/new_code?id=sdruskat_pyswh)
[![Docs build](https://readthedocs.org/projects/pyswh/badge/?version=latest)](https://pyswh.readthedocs.io/en/latest/?badge=latest)
[![codecov](https://codecov.io/gh/sdruskat/pyswh/branch/develop/graph/badge.svg?token=8IDQ3BXC4M)](https://codecov.io/gh/sdruskat/pyswh)
[![REUSE status](https://api.reuse.software/badge/github.com/sdruskat/pyswh)](https://api.reuse.software/info/github.com/sdruskat/pyswh)

## Getting started

Install `pyswh` via pip:

```bash
pip install pyswh
```

Include `pyswh` in your project by adding a respective dependency to your project, e.g.,

```bash
# requirements.txt
pyswh==0.1.0
```

```toml
# Poetry pyproject.toml
[tool.poetry.dependencies]
pyswh = "^0.1.0"
```

You can now use `pyswh`:

```python
from pyswh import swh
from pyswh import errors as swh_errors

try:
    swh.save('https://github.com/sdruskat/pyswh', False, 'SWH-API-AUTH-TOKEN')
except swh_errors.SwhSaveError as sse:
    raise sse
```

Refer to the [complete documentation](https://pyswh.readthedocs.io/en/latest/) to learn more about using `pyswh`.

## Set up for development

**Requirements:** Python >= 3.10.0.

1. Install [Poetry](https://python-poetry.org).

2. Clone the repository:

```bash
git clone git@github.com:sdruskat/pyswh.git
```

3. Create a virtual environment in `.venv`:
```bash
python3.10 -m venv .venv 
```

4. Activate the Poetry shell and install project:

```bash
poetry shell
poetry install
```

## Testing

`pyswh` uses `pytest` for testing. To run all tests, do:

```bash
poetry shell
poetry run pytest test/
```

## Building documentation locally

Initialize the Poetry virtual environment with `poetry shell`, go into the `docs/` folder and run `make html`.

## Licensing

See [`LICENSE.md`](LICENSE.md)

CHANGELOG.md
<!--
SPDX-FileCopyrightText: 2022 Stephan Druskat <pyswh@sdruskat.net>

SPDX-License-Identifier: CC0-1.0
-->

# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.0] - 2022-10-13

### Added
- Implementation of task to save code in the Software Heritage Archive

[Unreleased]: https://github.com/sdruskat/pyswh/compare/v0.1.0...HEAD
[0.1.0]: https://github.com/sdruskat/pyswh/releases/tag/v0.1.0


test/pyswh_test/test_swh.py
# SPDX-FileCopyrightText: 2022 Stephan Druskat <pyswh@sdruskat.net>
#
# SPDX-License-Identifier: MIT
import time
import logging

import pytest
import responses
import requests

from pyswh import swh
# from pyswh.errors import SwhSaveError


MOCK_SAVE_URL = 'https://archive.softwareheritage.org/api/1/origin/save/git/url/MOCK/'


@pytest.mark.parametrize('test_input, expected', [
    ('abc', 'abc/'),
    ('abc/', 'abc/'),
    ('https://www.domain.tld/dir', 'https://www.domain.tld/dir/'),
    ('https://www.domain.tld/dir/', 'https://www.domain.tld/dir/')
    ])
def test_prepare_url(test_input, expected):
    assert swh._prepare_url(test_input) == expected


@responses.activate
def test_back_off():

    current_epoch = int(time.time())
    responses.get('https://my.api/api/1/foobar',
                  headers={'X-RateLimit-Reset': str(current_epoch)},  # Rate limit will be "reset" in 2 secs.
                  status=200)
    response = requests.get('https://my.api/api/1/foobar')
    swh._back_off(response)
    assert int(time.time() - current_epoch) == 2  # 2 secs. added in MOT


@responses.activate
def test_check_rate_limit_pass(caplog):
    responses.get('https://archive.softwareheritage.org/api/1/ping/',
                  headers={'X-RateLimit-Remaining': str(1)})
    with caplog.at_level(logging.DEBUG):
        swh._check_rate_limit()
    assert len(caplog.records) == 0


@responses.activate
def test_check_rate_limit_429(caplog):
    current_epoch = int(time.time())
    responses.get('https://archive.softwareheritage.org/api/1/ping/', status=429,
                  headers={'X-RateLimit-Reset': str(current_epoch)})
    with caplog.at_level(logging.DEBUG):
        swh._check_rate_limit()
    assert int(time.time() - current_epoch) == 2  # 2 secs. added in MOT
    assert caplog.records[0].msg == 'Too many requests! Backing off.'
    assert 'Rate limit exceeded' in caplog.records[1].msg


@responses.activate
def test_check_rate_limit_rate_limit(caplog):
    current_epoch = int(time.time())
    responses.get('https://archive.softwareheritage.org/api/1/ping/',
                  headers={'X-RateLimit-Remaining': str(0),
                           'X-RateLimit-Reset': str(current_epoch)})
    with caplog.at_level(logging.DEBUG):
        swh._check_rate_limit()
    assert int(time.time() - current_epoch) == 2  # 2 secs. added in MOT
    assert caplog.records[0].msg == 'Rate limit exceeded. Backing off.'
    assert 'Rate limit exceeded' in caplog.records[1].msg


def test_build_request_url():
    assert swh._build_request_url('MOCK') == MOCK_SAVE_URL


@responses.activate
def test_request():
    responses.get('https://archive.softwareheritage.org/api/1/ping/',
                  headers={'X-RateLimit-Remaining': str(1)})
    responses.get(MOCK_SAVE_URL,
                  body='{"method": "GET"}', status=200,
                  content_type='application/json')
    responses.post(MOCK_SAVE_URL,
                   body='{"method": "POST"}', status=200,
                   content_type='application/json',
                   headers={'Authorization': 'Bearer xyz'})

    assert swh._request(swh._RequestMethod.POST, 'MOCK', None).content == b'{"method": "POST"}'
    assert swh._request(swh._RequestMethod.GET, 'MOCK', None).content == b'{"method": "GET"}'
    assert swh._request(swh._RequestMethod.POST, 'MOCK', 'xyz').headers['Authorization'] == 'Bearer xyz'


@responses.activate
def test_init_save_pass():
    responses.get('https://archive.softwareheritage.org/api/1/ping/',
                  headers={'X-RateLimit-Remaining': str(1)})
    responses.post(MOCK_SAVE_URL,
                   body='{"method": "POST"}', status=200,
                   content_type='application/json')
    assert swh._init_save('MOCK', None).content == b'{"method": "POST"}'


@responses.activate
def test_init_save_raise():
    with pytest.raises(swh.SwhSaveError,
                       match='Could not connect to the Software Heritage API. Are you connected to the internet?'):
        swh._init_save('MOCK', None)


@responses.activate
def test_check_save_progress_raise_connection():
    with pytest.raises(swh.SwhSaveError,
                       match='Could not connect to the Software Heritage API during progress check. '
                             'Are you connected to the internet?'):
        swh._check_save_progress('MOCK', None, '123')


@responses.activate
def test_check_save_progress_succeed(caplog):
    responses.get('https://archive.softwareheritage.org/api/1/ping/',
                  headers={'X-RateLimit-Remaining': str(1)})
    responses.get(MOCK_SAVE_URL,
                  headers={'X-RateLimit-Remaining': str(1)},
                  body='[{"loading_task_id": "123", "save_task_status": "succeeded", "visit_status": "full"}]',
                  status=200)
    with caplog.at_level(logging.DEBUG):
        swh._check_save_progress('MOCK', None, '123')
    assert caplog.records[1].msg == 'Saving MOCK has succeeded with visit status full!'


@responses.activate
def test_check_save_progress_fail(caplog):
    responses.get('https://archive.softwareheritage.org/api/1/ping/',
                  headers={'X-RateLimit-Remaining': str(1)})
    responses.get(MOCK_SAVE_URL,
                  headers={'X-RateLimit-Remaining': str(1)},
                  body='[{"loading_task_id": "123", "save_task_status": "failed", "visit_status": "full"}]',
                  status=200)
    with pytest.raises(swh.SwhSaveError,
                       match='Saving "MOCK" has failed with visit status "full"!'):
        swh._check_save_progress('MOCK', None, '123')


@responses.activate
def test_check_save_progress_wait(caplog):
    responses.get('https://archive.softwareheritage.org/api/1/ping/', headers={'X-RateLimit-Remaining': str(1)})
    responses.get(MOCK_SAVE_URL,
                  headers={'X-RateLimit-Remaining': str(1)},
                  body='[{"loading_task_id": "123", "save_task_status": "pending", "visit_status": "full"}]',
                  status=200)
    responses.get(MOCK_SAVE_URL,
                  headers={'X-RateLimit-Remaining': str(1)},
                  body='[{"loading_task_id": "123", "save_task_status": "succeeded", "visit_status": "full"}]',
                  status=200)
    with caplog.at_level(logging.DEBUG):
        swh._check_save_progress('MOCK', None, '123')
    assert caplog.records[1].msg == 'The save task for MOCK is pending. ' \
                                    'Waiting for 1 sec. before checking the status again.'
    assert caplog.records[3].msg == 'Saving MOCK has succeeded with visit status full!'


@responses.activate
def test_save_post_only():
    responses.get('https://archive.softwareheritage.org/api/1/ping/', headers={'X-RateLimit-Remaining': str(1)})
    responses.post(MOCK_SAVE_URL,
                   headers={'X-RateLimit-Remaining': str(1)},
                   body='{"loading_task_id": "123", "save_task_status": "succeeded", "visit_status": "full"}',
                   status=200)
    swh.save('MOCK', True, None)


@responses.activate
def test_save_400():
    responses.get('https://archive.softwareheritage.org/api/1/ping/', headers={'X-RateLimit-Remaining': str(1)})
    responses.post(MOCK_SAVE_URL, status=400, body=b'NO_URL')
    with pytest.raises(swh.SwhSaveError,
                       match=f'An invalid visit type or origin url has been provided.\n'
                             f'URL: MOCK\n{str(b"NO_URL")}'):
        swh.save('MOCK', False, None)


@responses.activate
def test_save_403():
    responses.get('https://archive.softwareheritage.org/api/1/ping/', headers={'X-RateLimit-Remaining': str(1)})
    responses.post(MOCK_SAVE_URL, status=403, body=b'BLACKLISTED')
    with pytest.raises(swh.SwhSaveError,
                       match=f'The provided origin url is blacklisted.\nURL: MOCK\n{str(b"BLACKLISTED")}'):
        swh.save('MOCK', False, None)


@responses.activate
def test_save_404():
    responses.get('https://archive.softwareheritage.org/api/1/ping/', headers={'X-RateLimit-Remaining': str(1)})
    responses.post(MOCK_SAVE_URL, status=404, body=b'NOT FOUND')
    with pytest.raises(swh.SwhSaveError,
                       match=f'No save requests have been found for a given origin.\nURL: MOCK\n{str(b"NOT FOUND")}'):
        swh.save('MOCK', False, None)


@responses.activate
def test_save_429(caplog):
    responses.get('https://archive.softwareheritage.org/api/1/ping/', headers={'X-RateLimit-Remaining': str(1)})
    responses.post(MOCK_SAVE_URL,
                   status=429,
                   headers={'X-RateLimit-Reset': str(int(time.time()) + 1)})
    responses.post(MOCK_SAVE_URL,
                   headers={'X-RateLimit-Remaining': str(1)},
                   body='{"loading_task_id": "123", "save_task_status": "succeeded", "visit_status": "full",'
                        '"origin_url": "MOCK", "save_request_status": "accepted"}',
                   status=200)
    responses.get(MOCK_SAVE_URL,
                  headers={'X-RateLimit-Remaining': str(1)},
                  body='{"loading_task_id": "123", "save_task_status": "succeeded", "visit_status": "full",'
                       '"origin_url": "MOCK", "save_request_status": "accepted"}',
                  status=200)
    with caplog.at_level(logging.DEBUG):
        swh.save('MOCK', False, None)
    assert caplog.records[1].msg == 'Rate limit exceeded. Waiting 3 seconds before retrying.'
    assert caplog.records[4].msg == 'Saving MOCK has succeeded with visit status full!'


@responses.activate
def test_save_unknown_status(caplog):
    responses.get('https://archive.softwareheritage.org/api/1/ping/', headers={'X-RateLimit-Remaining': str(1)})
    responses.post(MOCK_SAVE_URL, status=410, body=b'GONE')
    with pytest.raises(swh.SwhSaveError,
                       match='The status of the API response is unknown. '
                             'Please open a new issue reporting this at https://github.com/sdruskat/pyswh/issues. '
                             'Status code: 410'):
        swh.save('MOCK', False, None)


@responses.activate
def test_save_succeed(caplog):
    responses.get('https://archive.softwareheritage.org/api/1/ping/', headers={'X-RateLimit-Remaining': str(1)})
    responses.post(MOCK_SAVE_URL,
                   headers={'X-RateLimit-Remaining': str(1)},
                   body='{"loading_task_id": "123", "save_task_status": "pending", "visit_status": "full", '
                        '"origin_url": "MOCK", "save_request_status": "pending"}',
                   status=200)
    responses.get(MOCK_SAVE_URL,
                  headers={'X-RateLimit-Remaining': str(1)},
                  body='[{"loading_task_id": "123", "save_task_status": "succeeded", "visit_status": "full",'
                       '"origin_url": "MOCK", "save_request_status": "accepted"}]',
                  status=200)
    with caplog.at_level(logging.DEBUG):
        swh.save('MOCK', False, None)
    assert caplog.records[1].msg == 'The request to save MOCK is still pending. ' \
                                    'Waiting for 1 sec. before checking the status again.'
    assert caplog.records[4].msg == 'Saving MOCK has succeeded with visit status full!'
    assert caplog.records[6].msg == 'Saving MOCK has succeeded with visit status full!'


@responses.activate
def test_auth_token_header_log(caplog):
    responses.get('https://archive.softwareheritage.org/api/1/ping/', headers={'X-RateLimit-Remaining': str(1)})
    responses.post(MOCK_SAVE_URL,
                   headers={'X-RateLimit-Remaining': str(1)},
                   body='{"loading_task_id": "123", "save_task_status": "pending", "visit_status": "full", '
                        '"origin_url": "MOCK", "save_request_status": "pending"}',
                   status=200)
    with caplog.at_level(logging.DEBUG):
        swh.save('MOCK', True, 'xyz')
    assert caplog.records[0].msg == 'Making authenticated requests (authorization token).'


@responses.activate
def test_get_current_result_pass():
    responses.get(MOCK_SAVE_URL,
                  headers={'X-RateLimit-Remaining': str(1)},
                  body='[{"loading_task_id": "123", "msg": "HIT"},'
                       '{"loading_task_id": "122", "msg": "MISS"}]',
                  status=200)
    response_json = requests.get(MOCK_SAVE_URL).json()
    assert swh._get_current_result(response_json, '123')['msg'] == 'HIT'


@responses.activate
def test_get_current_result_raise():
    responses.get(MOCK_SAVE_URL,
                  headers={'X-RateLimit-Remaining': str(1)},
                  body='[{"loading_task_id": "123", "msg": "HIT", "origin_url": "MOCK"},'
                       '{"loading_task_id": "122", "msg": "MISS", "origin_url": "MOCK"}]',
                  status=200)
    response_json = requests.get(MOCK_SAVE_URL).json()
    with pytest.raises(swh.SwhSaveError,
                       match='Failed to retrieve the save task for MOCK.\n'
                             r'Full response: \[\{.+\}\]'):
        swh._get_current_result(response_json, '124')


@responses.activate
def test_check_status_rejected():
    responses.get(MOCK_SAVE_URL,
                  headers={'X-RateLimit-Remaining': str(1)},
                  body='{"loading_task_id": "123", "save_task_status": "succeeded", "visit_status": "full",'
                       '"origin_url": "MOCK", "save_request_status": "rejected", "note": "MISS"}',
                  status=200)
    response = requests.get(MOCK_SAVE_URL)
    with pytest.raises(swh.SwhSaveError,
                       match='The request to save MOCK has been rejected:\nNotes: MISS\n'
                             r'Full response: b.*'):
        swh._check_status(response, None, '123')


test/pyswh_test/test_errors.py
# SPDX-FileCopyrightText: 2022 Stephan Druskat <pyswh@sdruskat.net>
#
# SPDX-License-Identifier: MIT

import pytest

from pyswh import errors


@pytest.fixture()
def swh_save_error():
    return errors.SwhSaveError('I am an error.')


def test_swh_save_error(swh_save_error):
    assert type(swh_save_error) is errors.SwhSaveError
    assert type(swh_save_error) is not Exception


src/pyswh/swh.py
# SPDX-FileCopyrightText: 2022 Stephan Druskat <pyswh@sdruskat.net>
#
# SPDX-License-Identifier: MIT

from enum import Enum
import logging
import time
import typing as t

import requests
from requests import exceptions as request_exceptions

from pyswh.errors import SwhSaveError

_API_ROOT_URL = 'https://archive.softwareheritage.org/api/1/'
_API_ENDPOINT_SAVE = 'origin/save/'
_API_URL_PATH = '/url/'
_visit_type = 'git'  # TODO Add bzr, hg, svn

_log = logging.getLogger(__name__)


class _RequestMethod(Enum):
    """
    An enum representing API request methods, such as `post` and `get`.
    """
    POST = 'POST'
    GET = 'GET'


def _build_request_url(origin_url: str) -> str:
    """
    Constructs a valid request URL to use with the Software Heritage API from its parts.

    :param str origin_url: The URL for the origin source code repository that should be saved.
    :return: A valid request URL
    :rtype: str
    """
    return _prepare_url(_API_ROOT_URL + _API_ENDPOINT_SAVE + _visit_type + _API_URL_PATH + origin_url)


def _check_rate_limit():
    """
    Pings the SWH API to receive a response with rate limit information in the header,
    and triggers a backoff if the rate limit is exceeded, or an HTTP status code 429 (Too many requests) is encountered.
    :return:
    """
    response = requests.get('https://archive.softwareheritage.org/api/1/ping/')
    if response.status_code == 429:
        _log.info('Too many requests! Backing off.')
        _back_off(response)
        return

    if int(response.headers['X-RateLimit-Remaining']) > 0:
        return
    else:
        _log.info('Rate limit exceeded. Backing off.')
        _back_off(response)


def _request(method: _RequestMethod, origin_url: str, auth_token: str) -> requests.Response:
    """
    Makes a rate limit-safe request to the SWH API and returns the :py:class:`requests.Response`.

    :param _RequestMethod method: The request method to use for the request.
    :param str origin_url: The origin URL to use for construction of the request URL.
    :param str auth_token: An optional SWH auth token.
    :return: The response returned for the request.
    :rtype: requests.Response
    """
    _check_rate_limit()
    request_url = _build_request_url(origin_url)
    headers = {'Accept': 'application/json'}
    if auth_token:
        _log.debug('Making authenticated requests (authorization token).')
        headers['Authorization'] = f'Bearer {auth_token}'
    else:
        _log.debug('Making anonymous requests.')
    if method is _RequestMethod.POST:
        return requests.post(request_url, headers=headers)
    elif method is _RequestMethod.GET:
        return requests.get(request_url, headers=headers)


def _init_save(origin_url: str, auth_token: str) -> requests.Response:
    """
    Requests the initial save action in the Software Heritage API.

    :param str origin_url: The URL for the origin source code repository to save in the Software Heritage Archive.
    :param str auth_token: An optional SWH auth token.
    :return: The response returned for the request.
    :rtype: requests.Response
    :raises SwhSaveError: if no connection to the internet exists.
    """
    """"""
    try:
        return _request(_RequestMethod.POST, origin_url, auth_token)
    except request_exceptions.ConnectionError:
        raise SwhSaveError('Could not connect to the Software Heritage API. Are you connected to the internet?')


def _check_save_progress(origin_url: str, auth_token: str, task_id: str):
    """
    Checks on the progress of the save action, and reports its results.

    :param str origin_url: The URL for the origin source code repository to save in the Software Heritage Archive.
    :param str auth_token: An optional SWH auth token.
    :param str task_id: The task id of the save task, provided by the SWH API.
    :raises SwhSaveError: if no connection to the internet exists, or if the save action was unsuccessful.
    """
    try:
        response = _request(_RequestMethod.GET, origin_url, auth_token)
    except request_exceptions.ConnectionError:
        raise SwhSaveError('Could not connect to the Software Heritage API during progress check. '
                           'Are you connected to the internet?')

    response_json = response.json()

    if type(response_json) == list:
        response_json = _get_current_result(response_json, task_id)

    save_status = response_json['save_task_status']
    if save_status == 'failed':
        raise SwhSaveError(f'Saving "{origin_url}" has failed with visit status "{response_json["visit_status"]}"!'
                           f'\nFull response: {response.text}')
    elif save_status == 'succeeded':
        _log.info(f'Saving {origin_url} has succeeded with visit status {response_json["visit_status"]}!')
    else:  # One of not created, not yet scheduled, scheduled
        _log.info(f'The save task for {origin_url} is {save_status}. '
                  f'Waiting for 1 sec. before checking the status again.')
        time.sleep(1)
        _check_save_progress(origin_url, auth_token, task_id)


def _get_current_result(response_json: t.Any, task_id: str) -> t.Any:
    """
    Retrieves the current result from a list of save results that the SWH API returns when `get`ting the save action
    with the provided task id.

    :param t.Any response_json: A list of request results, i.e., JSON objects.
    :param str task_id: The identifier of the current save task.
    :return: The JSON response object for the current task id.
    :rtype: t.Any
    :raises SwhSaveError: if the response with the current task id cannot be found in the list of responses.
    """
    """Returns the object with the provided task id from a list of objects."""
    for obj in response_json:
        if obj['loading_task_id'] == task_id:
            return obj
    # If we're still here, the task ID could not be found
    raise SwhSaveError(f'Failed to retrieve the save task for {response_json[0]["origin_url"]}.\n'
                       f'Full response: {response_json}')


def _check_status(response: requests.Response, auth_token: str, task_id: str):
    """
    Checks the status of the save action as reported by the :py:class:`requests.Response`.

    :param requests.Response response: The response of the initial save request.
    :param str auth_token: An optional SWH auth token.
    :param str task_id: The task id of the current save task.
    :raises SwhSaveError: if the save request has been rejected.
    """

    # First, check the overall requests status (accepted, rejected, pending)
    response_json = response.json()
    if type(response_json) == list:
        response_json = _get_current_result(response_json, task_id)
    origin_url = response_json['origin_url']
    request_status = response_json['save_request_status']
    if request_status == 'pending':
        # Wait
        _log.info(f'The request to save {origin_url} is still pending. '
                  f'Waiting for 1 sec. before checking the status again.')
        time.sleep(1)
        retry_response = _request(_RequestMethod.GET, origin_url, auth_token)
        _check_status(retry_response, auth_token, task_id)
    elif request_status == 'rejected':
        raise SwhSaveError(f'The request to save {origin_url} has been rejected:\n'
                           f'Notes: {response_json["note"]}\nFull response: {response.content}')

    # Request status is accepted, check for save progress
    _check_save_progress(origin_url, auth_token, task_id)


def _prepare_url(origin_url: str) -> str:
    """
    Prepares the request URL for a request against the SWH API.

    The origin URL for the save request to the Software Heritage Archive API needs to end with a slash.
    Therefore, if there is no slash at the end of the provided origin URL, add it.

    :return the origin URL ending with a slash"""
    if origin_url.endswith('/'):
        return origin_url
    else:
        return origin_url + '/'


def _back_off(response: requests.Response):
    """
    Backs off from making further API calls for the amount of time that is the difference between
    the current epoch and the epoch when the rate limit will be reset.

    :param requests.Response response: The response from which to extract the rate limit reset time.
    """
    reset_time = int(response.headers['X-RateLimit-Reset'])
    now_time = int(time.time())
    sleep_time = reset_time - now_time + 2
    _log.info(f'Rate limit exceeded. Waiting {sleep_time} seconds before retrying.')
    time.sleep(sleep_time)  # Wait until the reset time, and an extra 2 seconds to be on the safe side.


def save(origin_url: str, post_only: bool, auth_token: str):
    """
    Attempts to save code in the Software Heritage Archive.

    This method wraps the `/api/1/origin/save/ <https://archive.softwareheritage.org/1/origin/save/doc/>`_ endpoint.

    :param str origin_url: The URL of the origin (source code repository) that should be saved in the archive.
    :param bool post_only: Whether the URL should simply be posted to the API and return,
        without checking for the success of the save operation.
    :param str auth_token: An optional Software Heritage API authentication token.
    :raises SwhSaveError: if an error occurred during the save task, or if the save task was unsuccessful.
    """
    if post_only:
        _init_save(origin_url, auth_token)
    else:
        init_response = _init_save(origin_url, auth_token)
        status = init_response.status_code
        if status == 200:
            # The API promises exactly one object as content of the POST response, so we can safely get the task ID
            task_id = init_response.json()['loading_task_id']
            _check_status(init_response, auth_token, task_id)
        elif status == 400:
            raise SwhSaveError(f'An invalid visit type or origin url has been provided.\n'
                               f'URL: {origin_url}\n'
                               f'{init_response.content}')
        elif status == 403:
            raise SwhSaveError(f'The provided origin url is blacklisted.'
                               f'\nURL: {origin_url}'
                               f'\n{init_response.content}')
        elif status == 404:
            raise SwhSaveError(f'No save requests have been found for a given origin.'
                               f'\nURL: {origin_url}'
                               f'\n{init_response.content}')
        elif status == 429:  # Too many requests
            _back_off(init_response)
            save(origin_url, False, auth_token)
        else:
            raise SwhSaveError(f'The status of the API response is unknown. '
                               f'Please open a new issue reporting this at https://github.com/sdruskat/pyswh/issues. '
                               f'Status code: {status}')
